### ISSUE TITLE
sim: add support for job cancellation and node exclusion

### ISSUE THREAD

**Issue by SteVwonder** [flux-framework/flux-sched#314]
Currently the simulator only supports job submissions.  It would be nice to also support job cancellations.  For example, the scenario where the job shows up in a SLURM trace but gets cancelled before it actually runs.  I believe that accounts for approx 30% of jobs at LLNL.

Given the recent addition of node exclusion (https://github.com/flux-framework/flux-sched/pull/305), it would be cool to add support for that in the simulation as well.  For example, we could replicate the effects of DATs and maintenance downtime.

This two things are functionality that were inquired about by one of Martin's students, Dai Yang, for his research.

**Comment by dongahn** [comment]
Would it be as simple as register the two request handler with the emulator code or will this be a bit more involved?

**Comment by dongahn** [comment]
Just to clarify:

> For the job cancellation, I was thinking we can modify the submit module to emit not only submissions but also cancellations (assuming the cancellation is known a priori).

cancel command is currently handled by `wreck`.  Any reason we want this in the submit? Or are you talking about the emulator's submit module?


>  In the case described by Dai Yang, I believe they want to analyze the system over the course of the simulation and create job failures/cancellations "at runtime" based on specific system criteria. In that scenario, I guess the sim_exec module will need extending. Although maybe we leave that as a separate branch and out of the master branch. I'm not sure how generalizable that scenario is. If it has a use-case outside of this specific research goal, then I think we include it.

Interesting. How about designing general job failure emulation interface to enable research like this. Seems we will need something like this going forward as well.

> For node exclusion, I am not sure exactly how we want to solve this problem. My initial thought is that node failures/drains/exclusions/inclusions that are known a priori can be input into the simulation with a separate input file (containing which nodes should go down and when) and handled either through the submit module or a new module. The module reads the node events in from the file and emits the events at the appropriate simulation time.

Similar, how about general node failure interface?

**Comment by SteVwonder** [comment]
> cancel command is currently handled by wreck. Any reason we want this in the submit? Or are you talking about the emulator's submit module?

Sorry for the confusion; you are correct. I was referring to the sim submit module.

> How about designing general job failure emulation interface to enable research like this. Seems we will need something like this going forward as well.

Sure.  I can think more deeply about this and the node failure emulation.  I think Dai's research will be a good driver & use-case for these features.

A related point: My understanding is that the scheduler currently does NOT differentiate between jobs that successfully exit (i.e., exit code == 0), jobs that "fail" (i.e., exit code != 0), and jobs that are killed for exceeding their walltime.  Inspired by the behavior of others schedulers, we might want to expose these differences to the user in commands like `flux wreck ls` as well as in features like job dependencies (e.g., only run this job after the successful completion of job 135).  I do not know if other schedulers support this, but it would also be cool to support resubmission and user-provided callbacks when a job fails or is killed.

**Comment by dongahn** [comment]
> Sure. I can think more deeply about this and the node failure emulation. I think Dai's research will be a good driver & use-case for these features.

What may make sense here is for us to design/implement a generic interface and researchers like Dai use that interface to emulate failure patterns they want.

> A related point: My understanding is that the scheduler currently does NOT differentiate between jobs that successfully exit (i.e., exit code == 0), jobs that "fail" (i.e., exit code != 0), and jobs that are killed for exceeding their walltime. Inspired by the behavior of others schedulers, we might want to expose these differences to the user in commands like flux wreck ls as well as in features like job dependencies (e.g., only run this job after the successful completion of job 135). I do not know if other schedulers support this, but it would also be cool to support resubmission and user-provided callbacks when a job fails or is killed.

Good points. I think there are three parts to this.

1. Can our future remote execution service harvest application failures like this and generate appropriate info (e.g., by emitting more refined job events.) I think this would be a good topic when we get to create job state RFC. @grondo or @garlick , do we plan to capture application level failures like this in our design of the wreck replacement? 

2. Once the information is propagated, we have to pin down on what the scheduler needs to do. Job dependency is one of the things that can make use of such information and we will need to pin down on a few powerful use cases and factor that into the design.

3. Of course, we need to flesh out the details on the user-facing interface. How users will express their dependencies. I was talking to @trws about user facing interface for resource request and we should augment that discussion to cover job dependencies...

**Comment by grondo** [comment]
> 1. Can our future remote execution service harvest application failures like this and generate appropriate info (e.g., by emitting more refined job events.) I think this would be a good topic when we get to create job state RFC. @grondo or @garlick , do we plan to capture application level failures like this in our design of the wreck replacement?

See above. We're already collecting and storing each task exit status in the job record, from which the mode of application failure for non-zero exit status can be derived. I was assuming we'd continue to do this for the wreck replacement.

I'm skeptical of more refined job exit status since that means anything waiting on a job to be "complete" will have a growing list of refined states to wait on. I would prefer using the UNIX process model and when all tasks have exited and IO flushed the job is "complete", at which point you can check the exit status (e.g. `waitpid`) and deduce the exit code, whether tasks were signaled and decide if the job "failed" or was "killed" depending on which tool is doing the reporting. It would be nice to have one or two terminal states: "completed" (all tasks successfully called exec() and exited) or "failed" (job failed to even run)

On the other hand, it might be useful to have events for job "timeout", "killed" or "failed", but those could be generated by a different job postprocessing service instead of being built-in job states.

I admit I might be oversimplifying so its all up for discussion, IMO.

However I'd hate to be in the same boat as some other resource managers with a job "state" plus a bunch of flags! (maybe that is just me though)

**Comment by trws** [comment]
On 17 Apr 2018, at 13:42, Mark Grondona wrote:

> However I'd hate to be in the same boat as some other resource 
> managers with a job "state" plus a bunch of flags! (maybe that is just 
> me though)

What do you mean here? I was thinking from the rest of your comment that 
a good solution might be to have an extra field in the “complete” 
state event that provides context, exit code some other small stuff that 
would otherwise have to be mined from the KVS with subsequent messages.  
That sounds a bit like what you’re saying you’re against here 
though.

**Comment by dongahn** [comment]
> I'm skeptical of more refined job exit status since that means anything waiting on a job to be "complete" will have a growing list of refined states to wait on. I would prefer using the UNIX process model and when all tasks have exited and IO flushed the job is "complete", at which point you can check the exit status (e.g. waitpid) and deduce the exit code, whether tasks were signaled and decide if the job "failed" or was "killed" depending on which tool is doing the reporting.

@grondo: Actually I like this model if this can be done as lightweight as possible.

**Comment by grondo** [comment]
> What do you mean here? I was thinking from the rest of your comment that a good solution might be to have an extra field in the “complete” state event that provides context, exit code some other small stuff that would otherwise have to be mined from the KVS with subsequent messages. That sounds a bit like what you’re saying you’re against here though.

No, I was thinking about this kind of thing:
```C
#define IS_JOB_FINISHED(_X)             \
        ((_X->job_state & JOB_STATE_BASE) >  JOB_SUSPENDED)
#define IS_JOB_COMPLETED(_X)            \
        (IS_JOB_FINISHED(_X) && ((_X->job_state & JOB_COMPLETING) == 0))
```

I guess it isn't so bad, but it would be nice if there was one "complete" state that event listeners could use to tell when a job was "finished"-- then another action to determine why it finished, if necessary.

**Comment by trws** [comment]
That makes sense to me.  Having something written to the kvs and/or included with that event/state would be good though.  While it isn’t complicated to interpret return codes, doing things like checking for whether any were signaled in external tools can be a bit cumbersome, so interpreting it into a sub-state that can just be read would ease things (this actually came up in the splash workflow tool, but it was easier to work around it than try and add something to flux for that timeframe).

On Apr 17, 2018, at 2:12 PM, Mark Grondona <notifications@github.com<mailto:notifications@github.com>> wrote:


What do you mean here? I was thinking from the rest of your comment that a good solution might be to have an extra field in the “complete” state event that provides context, exit code some other small stuff that would otherwise have to be mined from the KVS with subsequent messages. That sounds a bit like what you’re saying you’re against here though.

No, I was thinking about this kind of thing:

#define IS_JOB_FINISHED(_X)             \
        ((_X->job_state & JOB_STATE_BASE) >  JOB_SUSPENDED)
#define IS_JOB_COMPLETED(_X)            \
        (IS_JOB_FINISHED(_X) && ((_X->job_state & JOB_COMPLETING) == 0))

I guess it isn't so bad, but it would be nice if there was one "complete" state that event listeners could use to tell when a job was "finished"-- then another action to determine why it finished, if necessary.

**Comment by grondo** [comment]
closed by #440, or do we want to move this to flux-core? Feel free to reopen and move if so
